Rewrite of password reset testing (#743)
OWASP/wstg: document/4-Web_Application_Security_Testing/04-Authentication_Testing/09-Testing_for_Weak_Password_Change_or_Reset_Functionalities.md

## Summary

For any application that requires the user to authenticate with a password, there must be a mechanism by which the user can regain access to their account if they forget their password. Although this can sometimes be a manual process that involves contacting the owner of the website or a support team, users are frequently allowed to carry out a self-service password reset, and to regain access to their account by providing some other evidence of their identity.

As this functionality provides a direct route to compromise the user's account, it is crucial that it is implemented securely.

## Test Objectives

- Determine whether the password change and reset functionality allows accounts to be compromised.

## How to Test

### Information Gathering

> The first step is to gather information about what mechanisms are available to allow the user to reset their password on the application. If there are multiple interfaces on the same site (such as a web interface, mobile application and API) then these should all be reviewed, in case they provide different functionality.

Reviewer: ```suggestion
The first step is to gather information about what mechanisms are available to allow the user to reset their password on the application. If there are multiple interfaces on the same site (such as a web interface, mobile application, and API) then these should all be reviewed, in case they provide different functionality.
```

Once this has been established, determine what information is required in order for a user to initiate a password reset. This is often the username or email address (both of which are easily guessable), but it could also be an internally generated user ID, which would be harder to obtain.

### General Concerns

Regardless of the specific methods used to reset passwords, there are a number of common areas that need to be considered:

- Is the password reset process weaker than the authentication process?

  The password reset process provides an alternative mechanism to access a user's account, and so should be at least as secure as the usual authentication process. However, it can often provide an easier way to compromise the account, especially if it uses weaker authentication factors such as security questions.

>   Additionally, the password reset process often bypasses the requirement to use Multi-Factor Authentication (MFA), which can substantially reduce the security of the application.

Author: Depending on the implementation, the lack of MFA here can effectively provide a bypass for MFA on the main login functionality - so I think it undermine more than just the the security of the password reset functionality itself.

>   Additionally, the password reset process often bypasses the requirement to use Multi-Factor Authentication (MFA), which can substantially reduce the security of the application.

Reviewer 2: Okay, fair point.

- Is the password securely generated?

  The password should be generated using a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to prevent password guessing or brute-force attacks. For a secure user-friendly experience, it should be generated using a secure passphrase-style approach (i.e, combining multiple words), rather than a string of random characters.

- Is the user's existing password sent to them?

  Rather than generating a new password for the user, some applications will send the user their existing password. This is a very insecure approach, as it exposes their current password over unencrypted email. Additionally, if the site is able to recover the existing password, this implies that passwords are either stored using reversible encryption, or (more likely) in unencrypted plain text, both of which represent a serious security weakness.

- Are the emails sent from a domain with anti-spoofing protection?

>   The domain should implement SPF, DKIM and DMARC to prevent attackers from spoofing emails from it, which could be used as part of a social engineering attack.

Reviewer: ```suggestion
  The domain should implement SPF, DKIM, and DMARC to prevent attackers from spoofing emails from it, which could be used as part of a social engineering attack.
```

  Emails are typically sent unencrypted, and in many cases the user's email account will not be protected by MFA. It may also be shared between multiple individuals, particularly in a corporate environment.

  Consider whether email-based password reset functionality is appropriate, based on the context of the application that is being tested.

### Email - Link Sent

In this model, the user is sent a link via email which contains token. They can then click this link, and are prompted to enter a new password on the site. This is the most common approach used for password reset, but is more complex to implement than the previously discussed approach. The key areas to test are:

- Does the link use HTTPS?

  If the token is sent over unencrypted HTTP, it may be possible for an attacker to intercept it.

- Can the link be used multiple times?

  Links should expire after they are used, otherwise they provide a persistent backdoor for the account.

- Does the link expire?

  Links should be time limited. Exactly how long is appropriate will depend on the site, but it should rarely be more than an hour.

- Is token sufficiently long and random?

>   The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. As such the tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to effectively impossible to guess (at least 128 bits / 32 hex characters).

Author: I think of "crack" as a term that should be used for offline attacks (i.e, against hashes), rather than an online attack like this.

Personally, I prefer to give some kind of indication of length here - are we happy that people will have a good idea of what is and isn't practical to attack? Given it's an online attack, 128 bits will never be feasible (and has a pretty big safety margin), so I think it's a safe recommendation to give.

Not sure what the WSTG does elsewhere though.

>   The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. As such the tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to effectively impossible to guess (at least 128 bits / 32 hex characters).

Reviewer 2: How about this? I’m wary of absolutes like “impossible” and “never” and see my crypto betters already recommending 256. I also want to avoid making crypto recommendations here in the testing guide, since we’re not focused on remediation.

``` suggestion
  The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. The tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long so that it is impractical to brute-force given widely-available computing power.
```

>   The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. As such the tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to effectively impossible to guess (at least 128 bits / 32 hex characters).

Author: If this was something that could be attacked offline (i.e, a hash or encryption) then I'd absolutely agree with you. Also 128 bits is considered unbreakable, in 99.9% of cases there's no reason not to use 256 bits, so I'd definitely recommend that.

However for a (random) password reset token, it would have to be an online attack against the server itself, so the level of security needed is much lower because it's limited by the number of requests that can be made, rather than by compute power. Even 40 bits (~1.1 trillion possible tokens) would be impractical for a online attack (~15 years at 1,000 requests/s).

If we don't want to include any actual numbers here that's fine - it could be something like:

> ...sufficiently long that is is impractical for an attacker to guess or brute-force.

But people may interpret that very differently. Six lower-alphanumeric characters (~31 bits) would be impractical for an attacker to brute-force, but I would make an issue of it if I saw an application using a token that length.

>   The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. As such the tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to effectively impossible to guess (at least 128 bits / 32 hex characters).

Reviewer 2: Hmm, I see your point. Okay, how about this then:

``` suggestion
  The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. The tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long that it is impractical for an attacker to guess or brute-force. At least 128 bits (or 32 hex characters) is a sufficient minimum to make such an online attack impractical.
```

>   The security of the process is entirely reliant on an attacker not being able to guess or brute-force a token. As such the tokens should be generated with a Cryptographically Secure Pseudo-Random Number Generator (CSPRNG), and should be sufficiently long to effectively impossible to guess (at least 128 bits / 32 hex characters).

Author: Seems reasonable.

  Tokens should never be generated based on known values, such as by taking the MD5 hash of the user's email with `md5($email)`, or using GUIDs which may use insecure PRNG functions, or may not even be random depending on the type.

>   An alternative approach to random tokens is to use a cryptographically signed token such as a JWT. In this case, the usual JWT checks should be carried out (is the signature verified, can the "nONe" algorithm be used, can the HMAC key be brute-forced, etc). See the [Testing for APIs](../../../Testing_for_APIs.md) guide for further information.

Reviewer: ```suggestion
  An alternative approach to random tokens is to use a cryptographically signed token such as a JWT. In this case, the usual JWT checks should be carried out (is the signature verified, can the "nONe" algorithm be used, can the HMAC key be brute-forced, etc). See the [Testing JSON Web Tokens](../../06-Session_Management_Testing/10-Testing_JSON_Web_Tokens.md) guide for further information.
```

>   An alternative approach to random tokens is to use a cryptographically signed token such as a JWT. In this case, the usual JWT checks should be carried out (is the signature verified, can the "nONe" algorithm be used, can the HMAC key be brute-forced, etc). See the [Testing JSON Web Tokens](../../06-Session_Management_Testing/10-Testing_JSON_Web_Tokens.md) guide for further information.

Reviewer: ```suggestion
  An alternative approach to random tokens is to use a cryptographically signed token such as a JWT. In this case, the usual JWT checks should be carried out (is the signature verified, can the "nONe" algorithm be used, can the HMAC key be brute-forced, etc). See the [Testing JSON Web Tokens](../06-Session_Management_Testing/10-Testing_JSON_Web_Tokens.md) guide for further information.
```

  Sometimes the password reset link may include a user ID as well as a token, such as `reset.php?userid=1&token=123456`. In this case, it may be possible to modify the `userid` parameter to reset other users' passwords.

- Can you inject a different host header?

  If the application trusts the value of the `Host` header and uses this to generate the password reset link, it may be possible to steal tokens by injecting a modified `Host` header into the request. See the [Testing for Host Header Injection](../07-Input_Validation_Testing/17-Testing_for_Host_Header_Injection.md) guide for further information.

- Is the link exposed to third parties?

  If the page that the user is taken to includes content from other parties (such as loading scripts from other domains), then the reset token in the URL may be exposed in the HTTP `Referer` header sent in these requests. The `Referrer-Policy` HTTP header can be used to protect against this, so check if one is defined for the page.

  Additionally, if the page includes any tracking, analytics or advertising scripts, the token will also be exposed to them.

- Are the emails sent from a domain with anti-spoofing protection?

>   The domain should implement SPF, DKIM and DMARC to prevent attackers from spoofing emails from it, which could be used as part of a social engineering attack.

Reviewer: ```suggestion
  The domain should implement SPF, DKIM, and DMARC to prevent attackers from spoofing emails from it, which could be used as part of a social engineering attack.
```

- Is email considered sufficiently secure?

  Emails are typically sent unencrypted, and in many cases the user's email account will not be protected by MFA. It may also be shared between multiple individuals, particularly in a corporate environment.

  Consider whether email-based password reset functionality is appropriate, based on the context of the application that is being tested.

### Tokens Sent Over SMS or Phone Call

Rather than sending a token in an email, an alternative approach is to send it via SMS or an automated phone call, which the user will then enter on the application. The key areas to test are:

- Is the token sufficiently long and random?

>   Tokens sent this way are typically shorted, as they are intended to be manually typed by the user, rather than being embedded in a link. It's fairly common for applications to use six numeric digits, which only provides ~20 bits of security (feasible for an online brute-force attack), rather than the 128+ bits that would be present in an emailed token.

Reviewer 2: ```suggestion
  Tokens sent this way are typically shorted, as they are intended to be manually typed by the user, rather than being embedded in a link. It's fairly common for applications to use six numeric digits, which only provides ~20 bits of security (feasible for an online brute-force attack), rather than the typically longer email token.
```

Not a guarantee.

>   Tokens sent this way are typically shorted, as they are intended to be manually typed by the user, rather than being embedded in a link. It's fairly common for applications to use six numeric digits, which only provides ~20 bits of security (feasible for an online brute-force attack), rather than the typically longer email token.

Reviewer: ```suggestion
  Tokens sent this way are typically shorter, as they are intended to be manually typed by the user, rather than being embedded in a link. It's fairly common for applications to use six numeric digits, which only provides ~20 bits of security (feasible for an online brute-force attack), rather than the typically longer email token.
```

  This makes it much more important that the password reset functionality is protected against brute-force attacks.

- Can the token be used multiple times?

  Tokens should be invalidated after they are used, otherwise they provide a persistent backdoor for the account.

> - Does the token expire?

Reviewer 2: ```suggestion
- Does the token expire if it remains unused?
```

>   As the tokens are more susceptible to brute-force attacks, a shorter expiration time should be implemented to limit the window available for an attacker to carry out an attack.

Reviewer 2: ```suggestion
  As the shorter tokens are more susceptible to brute-force attacks, a shorter expiration time should be implemented to limit the window available for an attacker to carry out an attack.
```

- Are appropriate rate limiting and restrictions in place?

  Sending an SMS or triggering an automated phone call to a user is significantly more disruptive than sending an email, and could be used to harass a user, or even carry out a denial of service attack against their phone. The application should implement rate limiting to prevent this.

  Additionally, SMS messages and phone calls often incur financial costs for the sending party. If an attacker is able to cause a large number of messages to be sent, this could result in significant costs for the website operator. This is especially true if they are sent to international or premium rate numbers. However, allowing international numbers may be a requirement of the application.

- Is SMS or a phone call considered sufficiently secure?

>   A variety of attacks have been demonstrated that would allow an attacker to effectively hijack SMS messages, there are conflicting views about whether SMS is sufficiently secure to be used as an authentication factor.

Reviewer 2: ```suggestion
  As a variety of attacks have been demonstrated that would allow an attacker to effectively hijack SMS messages, there are conflicting views about whether SMS is sufficiently secure to be used as an authentication factor.
```

Citation or references? Is there an OWASP project that covers this in more detail?

Author: I don't know if there are any OWASP specific resources, but I'll have a look and find a good reference for this.

Author: How about this for a reference?

https://www.ncsc.gov.uk/guidance/protecting-sms-messages-used-in-critical-business-processes#section_4

Reviewer: Yup that one fits well.

Author: Added in 2b7644c.

- When setting the password, can you specify the user ID?

  If the user ID is included in the password reset request and is not validated, it may be possible to modify it and change other users' passwords.

- Is the user required to re-authenticate?

  If a logged-in user tries to change their password, they should be asked to re-authenticate with their current password in order to protect against an attacker gaining temporary access to an unattended session. If the user has MFA enabled, then they would typically re-authenticate with that, rather than their password.

- Is the password change form vulnerable to CSRF?

  If the user isn't required to re-authenticate, then it may be possible to carry out a CSRF attack against the password reset form, allowing their account to be compromised. See the [Testing for Cross-Site Request Forgery](../06-Session_Management_Testing/05-Testing_for_Cross_Site_Request_Forgery.md) guide for further information.

- Is a strong and effective password policy applied?

>   The password policy should be consistent across the registration, password change and password reset functionality. See the [Testing for Weak Password Policy](07-Testing_for_Weak_Password_Policy.md) guide for further information.

Reviewer: ```suggestion
  The password policy should be consistent across the registration, password change, and password reset functionality. See the [Testing for Weak Password Policy](07-Testing_for_Weak_Password_Policy.md) guide for further information.
```
